### ISSUE TITLE
monitor code quality

### ISSUE THREAD

**Issue by infinitewarp** [cloudigrade/cloudigrade#12]
As a Doppler developer, I want to see metrics around code quality so I can know when the overall health of the code base needs attention.

Options to consider:
- https://codacy.com
- https://codebeat.co
- https://landscape.io
- https://codeclimate.com
- https://sonarcloud.io

Example of codacy on a personal repo of mine:
https://www.codacy.com/app/infinitewarp/trekipsum/dashboard

**Comment by kdelee** [comment]
As a user of your issue tracker, I think that the `story` label does not really fit this. Maybe y'all could decide on some label for infrastructure improvments.

**Comment by infinitewarp** [comment]
One strike against https://landscape.io is that it's been down for the last hour that I've been trying to access it, and they don't have a public external status page to see what's going on.

**Comment by infinitewarp** [comment]
My findings aren't great. Maybe we abandon this idea for now or accept it in a limited fashion.

Here are the integrations I was able to set up:

- https://codeclimate.com/github/cloudigrade/cloudigrade
- https://codebeat.co/projects/github-com-cloudigrade-cloudigrade-master
- https://app.codacy.com/app/infinitewarp/cloudigrade/dashboard

My experiences and thoughts with each of the services so far:

- **Codebeat** may be dead. 🙁  It was able to clone and set up our project, but it claims that every file has _zero lines_ of code and it found no issues with our code. Their presence on Twitter fell silent in Oct 2017.
- **Landscape** may be dead. 🙁  Their website is down more often than it's up, and I haven't been able to get our project set up. Their presence on Twitter fell silent in May 2017.
- **Codacy** is pretty nice, but it ties the project through someone's personal github account (see the URL above that contains my "infinitewarp" username) instead of just using the organization name, and it seems to have some problems with organization-based repos (although this may be ephemeral?). Codacy did find a few minor issues with our code. Codacy has quality settings that should cause GitHub pull requests to fail, ~but it did nothing for a new pull request I created; it didn't even appear in the list of checks.~ Correction: it took a while, but eventually (I checked back 2 hours later) I did see the pull request appear on Codacy, and its check was integrated in the PR. As hoped, Codacy did find problems in the new code and blocks merging on the PR.
- **Code Climate** appears to integrate with GitHub and claims to create issues and update pull requests. Code Climate found no issues with our code, and although it did integrate as a check with a pull request I created, it found no issues with the deliberately badly crafted code I pushed.
- **SonarCloud** does not automatically run quality checks against the code and requires you to run `sonar-scanner` locally and upload results periodically. It looks like SonarCloud is just an alternate reporting view for the traditional SonarQube. Because of this requirement, I didn't get far enough to see if or how well it integrates with GitHub. :confused:

In summary, it looks like Codacy _could be_ our best fit, but I'm wary about the delay it experienced in discovering my pull request. Under the hoods, Codacy is just a history tracker and graphical wrapper over a suite of existing code-checking tools (PyLint, Prospector, and Bandit). We may be better off shooting down the idea of integrating with a third-party service and instead just expand our tox and Travis-CI setup to include execution of those additional tools.

**Comment by ilyavw** [comment]
It doesn't look like any of those tools really offer us anything good without drawbacks. I feel pretty good about the current setup we have and wouldn't feel bad about parking this idea for now.

**Comment by infinitewarp** [comment]
Okay, let's just close this story and we will consider directly adding checks via PyLint, etc. later.
